Added contribution guidelines to repository (#1404)
kirbydesign/designsystem: .github/CONTRIBUTING.md

If you at any point have any questions please read: [I have a question related to Kirby](#i-have-a-question-related-to-kirby)

- [The contribution model](#the-contribution-model)
  - [I have a feature request](#i-have-a-feature-request)
  - [I have an enhancement](#i-have-an-enhancement)
  - [I would like to report a bug](#i-would-like-to-report-a-bug)
- [Setup environment and tooling for contribution](#setup-environment-and-tooling-for-contribution)
  - [Clone and install repository](#clone-and-install-repository)
  - [Node](#node)
  - [Tslint](#tslint)
  - [Prettier](#prettier)
- [Creating pull-requests](#creating-pull-requests)
  - [Before you submit](#before-you-submit)
  - [After you submit](#after-you-submit)
- [The good:](#the-good:)
  - [Issue](#issue)
  - [Self-review](#self-review)
  - [Pull-request](#pull-request)
  - [Test](#test)
- [I have a question related to Kirby](#i-have-a-question-related-to-kirby)
  - [Submit an inquiry request on github](#submit-an-inquiry-request-on-github)
  - [Ask in the kirby-guild slack channel](#ask-in-kirby-guild)

## The contribution model

> <!-- This section has the purpose of explaining why we do it like we do, in order to have people 

Reviewer: Good point @RasmusKjeldgaard  👍🏻

> <!-- This section has the purpose of explaining why we do it like we do, in order to have people 

Author: Good point - I totally neglected that when writing the outline! 😀 
The new component request part now has more focus on UX- refinement and -review. It's still quite WET as @jakobe pointed out, but I will attempt to make it more DRY.

symphatize with the process -->
- What are we trying to achieve with this contribution model
- General description of the contribution model

<!-- The following "I have/would" sections has the role of easily guiding the reader 
on how to do their specific task, and explain them what the process is for that task -->
### I have a feature request 

> - Submit issue using the feature request template

Reviewer: ```suggestion
- Submit [new issue](https://github.com/kirbydesign/designsystem/issues/new/choose) using the feature request template
```

Author: 5ab1170

  - please read "[the good: issue](#issue)"
- Contact kirby-guild to get it tech refined 
- If you would like to implement it:
  - please read the following before continuing: 
    - "[Setup environment and tooling for contribution](#setup-environment-and-tooling-for-contribution)" to get properly setup 
    - "[Creating pull-requests](#creating-pull-requests)" to get informed on how to most easily get your changes merged
  - assign yourself to the issue to signal to others you are working on it
  - if you at any point have questions while solving the issue please see: [I have a question related to Kirby](#i-have-a-question-related-to-kirby)

### I have an enhancement 

> - Submit issue using the enhancement request template

Reviewer: ```suggestion
- Submit [new issue](https://github.com/kirbydesign/designsystem/issues/new/choose) using the enhancement request template
```

Author: 5ab1170

  - please read "[the good: issue](#issue)"
- Contact kirby-guild to get it tech refined 
- If you would like to implement issue: 
  -please read: 
    - "[Setup environment and tooling for contribution](#setup-environment-and-tooling-for-contribution)" to get properly setup 
    - "[Creating pull-requests](#creating-pull-requests)" to get informed on how to most easily get your changes merged
  - assign yourself to the issue to signal to others you are working on it
  - if you at any point have questions while solving the issue please see: [I have a question related to Kirby](#i-have-a-question-related-to-kirby)

### I would like to report a bug 

> - Submit issue using the bug template  

Reviewer: ```suggestion
- Submit [new issue](https://github.com/kirbydesign/designsystem/issues/new/choose) using the bug template  
```

Author: 5ab1170

  - please read "[the good: issue](#issue)"
- Contact kirby-guild to get it verified 
  - we verify because it might be a feature not a bug 🐛
- If you would like to implement issue: 
  -please read: 
    - "[Setup environment and tooling for contribution](#setup-environment-and-tooling-for-contribution)" to get properly setup 
    - "[Creating pull-requests](#creating-pull-requests)" to get informed on how to most easily get your changes merged
  - assign yourself to the issue to signal to others you are working on it

>   - if you at any point have questions while solving the issue please see: [I have a question related to Kirby](#i-have-a-question-related-to-kirby)

Reviewer: On one hand it's nice to have everything contained within each type (comp/enhancement/bug) on the other hand it's pretty WET - could we merge and move `If you would like to implement...` below 3 types? 🤔

Author: Yeah tought that myself. I have tried to write out the feature request part (which I've now renamed to component request), and will try to write out both the bug and enhancement part - while trying to format it to be more dry. 
But it's a balance between "here is a nice ordered checklist you should just follow"  and what I guess we can call "section-hopping" :)

## Setup environment and tooling for contribution
<!-- The following section is responsible for teaching people how to get setup with 
the repo such that they can get coding -->

### Clone and install repository 
- how to clone and get going with the kirby repo

### Node 
- Which version should people use, and perhaps a recommendation on which tool people can use to manage versions. 

### Tslint 
- People should make sure to have it enabled to avoid unnecessary comments during code review

### Prettier
- It is a good idea to use format-on-save

## Creating pull-requests 
<!-- This section is responsible for guiding contributors on how they should submit pull-requests in a way that ensures the most swift process and teach them how to catch low-hanging issues there might come up in code-review. -->

### Before you submit  
- Make sure you have completed the steps in: [Setup tooling for contribution](#setup-tooling-for-contribution)
- Make sure you have written tests according to the guidelines described in [The good: test](#test)
- Do a self-review <!-- We could link to self-review in the templates and have people check it off? -->

### After you submit 

> - Post in Kirby-guild 

Author: We should probably have something about how to get access to kirby-guiod

Author: Totally up for changing the process to be more "repo oriented". It would cut out some steps for contributors and also feels more OSS as you point out. But this should probably be discussed in plenum 😄

- Code reviews are meant to align your code with existing code 

>   - While your implementation might be perfectly valid, we are aiming for uniformity across the codebase. 

Reviewer: Nice! 👌🏻

> - Don't commit and exit - make sure to stay responsive and come back to answer comments on your code review

Reviewer: ```suggestion
- Please don't fire & forget, i.e. commit & run - make sure to stay responsive and come back to answer comments on your code review :)
```

Author: 5ab1170

> ## The good:

Reviewer 2: Maybe add something like 
### Commit
 - Focus on one thing per commit
 - Keep changes as small as possible
 - Only commit related functionality for each commit
 - Commit often
 - Strive for each commit to compile/build and pass tests

Reviewer 2: And maybe also add that the CI pipeline handles combining each commit to a merge commit containing all commit messages

Author: e15cb28 😄

<!-- Ideally we should be able to refer to these sections if people break the guidelines in their issue or pull-requests, to make sure we are all preaching the same guidelines -->

### Issue 
- General properties we are looking for and how to achieve them 
  - Readability 
  - Easily understandable 
    - No internal lingo 
  - Proof-read
  - Self-contained
    - Screenshots if necessary
    <!-- I think the following point is important if Kirby is to be spread out amongst several teams -->
    - Don't assume people have knowledge outside of Kirby - contributors might come from many different projects 

### Self-review 
- Don't change files that is not related to your pull-request
  - for example don't reorder imports in files that is not directly related to the issue at hand
- Check for unused imports
- Other self review points go here

### Pull-request
- How is a pull-request that is easily reviewed constructed?

### Test
- How should testing be done 

## I have a question related to Kirby
<!-- This section is responsible for teaching contributors how to get help with Kirby, we should probably link to this section from the readme.md -->

> We have **Kirby Open Office** every Tuesday and Thursday between 9:15 to 10:15. During these timeslots we will answer questions asked in the @kirby-design/kirby-guild slack channel and inquiry requests submitted on github. You are always welcome to ask questions outside these timeslots and we will answer it when we got time. 

Reviewer: Miiiiiight be a bit too specific for an OSS repo...
Maybe just something about that we have are continuously watching our repo for new issues, inquires and PRs...

> **Do not send direct messages to members of team-kirby:** This is not because we do not want to talk to you - we however prefer questions to be asked in public, such that everyone can benefit from the answers. 

Reviewer 3: ```Do not send direct messages to members of team-kirby:** This is not because we do not want to talk to you - we do however prefer questions to be asked in public, so that everyone can benefit from the answers.
```

### After you submit 
- Post in Kirby-guild 
- Code reviews are meant to align your code with existing code 
  - While your implementation might be perfectly valid, we are aiming for uniformity across the codebase. 
- Don't commit and exit - make sure to stay responsive and come back to answer comments on your code review

## The good:
<!-- The following sections have the responsibility of explaining contributors our best pratices such that they can more easily have their changes implemented in kirby --> 
<!-- Ideally we should be able to refer to these sections if people break the guidelines in their issue or pull-requests, to make sure we are all preaching the same guidelines -->

### Issue 
- General properties we are looking for and how to achieve them 
  - Readability 
  - Easily understandable 
    - No internal lingo 
  - Proof-read
  - Self-contained
    - Screenshots if necessary
    <!-- I think the following point is important if Kirby is to be spread out amongst several teams -->
    - Don't assume people have knowledge outside of Kirby - contributors might come from many different projects 

### Commit

> - Focus on one thing per commit 

Reviewer: Use Gitmoji's when possible! 😂

Author: 5ab1170

### Issue 
- General properties we are looking for and how to achieve them 
  - Readability 
  - Easily understandable 
    - No internal lingo 
  - Proof-read
  - Self-contained
    - Screenshots if necessary
    <!-- I think the following point is important if Kirby is to be spread out amongst several teams -->
    - Don't assume people have knowledge outside of Kirby - contributors might come from many different projects 

### Commit
- Focus on one thing per commit 
- Keep changes as small as possible 
- Only commit related functinoality for each commit 
- Commit often
- Strive for each commit to compile/build and pass tests 
- Use [gitmoji](https://gitmoji.dev) when possible - point to VSCode extension and CLI tool
- sidenote: when merging the CI pipeline combines each commit to a merge commit containing all commit messages

### Self-review 
- Don't change files that is not related to your pull-request
  - for example don't reorder imports in files that is not directly related to the issue at hand
- Check for unused imports
- Other self review points go here

### Pull-request
- How is a pull-request that is easily reviewed constructed?

### Test
- How should testing be done 

> ## I have a question related to Kirby

Author: This section should probably be moved to a SUPPORT.md file (https://github.com/kirbydesign/designsystem/issues/1462)

### After you submit 
- Post in Kirby-guild 
- Code reviews are meant to align your code with existing code 
  - While your implementation might be perfectly valid, we are aiming for uniformity across the codebase. 
- Please don't fire & forget, i.e. commit & run - make sure to stay responsive and come back to answer comments on your code review :)
## The good:
<!-- The following sections have the responsibility of explaining contributors our best pratices such that they can more easily have their changes implemented in kirby --> 
<!-- Ideally we should be able to refer to these sections if people break the guidelines in their issue or pull-requests, to make sure we are all preaching the same guidelines -->

### Issue 
- General properties we are looking for and how to achieve them 
  - Readability 
  - Easily understandable 
    - No internal lingo 
  - Proof-read
  - Self-contained
    - Screenshots if necessary
    <!-- I think the following point is important if Kirby is to be spread out amongst several teams -->
    - Don't assume people have knowledge outside of Kirby - contributors might come from many different projects 

> ### Branch 

Author: https://github.com/kirbydesign/designsystem/pull/1464#discussion_r613996711

First of all - thank you for your interest in contributing to Kirby! 🥳
<!-- Early on show that we are here to help if necessary and to educate people in how to get help -->
If you are looking for help or at any point have questions, check out "[Help!: i have a question related to kirby](SUPPORT.md)". 

> 

Author: When everything else is done, we should throw in a "How to use this document" section here.

If you need to have multiple versions of node installed then a version manager such as [NVM](https://github.com/nvm-sh/nvm) can be used. 

## TSlint 

<!-- People should make sure to have it enabled to avoid unnecessary comments during code review-->
**Note:** Work is being done to migrate from TSLint to ESLint (see issue: [#1317](https://github.com/kirbydesign/designsystem/issues/1317)). 

TSLint is used for linting which helps keep up the quality of the submitted code for Kirby. 

Scripts for running TSLint can be found in `package.json` . We however recommend that you install a TSLint plugin for your editor. It can be useful to enable "fix on save"-functionality such that simple errors are fixed automatically. 

**Useful resources:**

* [TSLint extension for VSCode](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin).
* [Enabling "fix on save" in VSCode](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin#auto-fix-and-auto-fix-on-save).

> ### VSCode "workspace library execution"-issue

Author: Is this still an issue?

Reviewer 2: Yes, I just got it this morning.

Reviewer 2: Should we add `.vscode/settings.json` to `.gitignore`?

<!-- We should probably link to this file in #kirby-guild -->
First of all - thank you for your interest in contributing to Kirby! 🥳
<!-- Early on show that we are here to help if necessary and to educate people in how to get help -->
If you at any point have any questions please read: [I have a question related to Kirby](#i-have-a-question-related-to-kirby)

**Beware:** We reserve the right to close issues and pull requests that does not follow the steps outlined in [the contribution model](#the-contribution-model). This is not because we want to be mean or micro-manage the way things are done - but a project of this size requires some consistency in the way we cooperate in order to be manageable. 

> * [Introduction to the contribution model](#introduction-to-the-contribution-model)

Author: Consider document outline and heading levels - also we do not have any H1/top level heading 😄

Author: Done

You can contribute yourself! And no matter your role, you can probably contribute with something. May it be a helpful comment or a bug report.

We think that is neat. 

There is still need for managing Kirby however. We need to make sure that the project is moving in the right direction. That submitted code is: high quality, not duplicating anything and consistent with the rest of the codebase.
To do so a series of steps has been created that an issue must go through. All of these aim at making sure that the process of cooperating from idea to implementation is smooth for contributors. With as little time wasted as possible.

> Simplicity—the art of maximizing the amount of work not done—is essential.
>  
> _[Principles behind the Agile Manifesto](https://agilemanifesto.org/principles.html)_

### The process of contributing

<!-- General description of the contribution model -->

In broad strokes the process is:

1. Submit an issue
2. Get it refined
3. Implement a solution
4. Get the solution reviewed and approved

> Once a pull-request has been approved it will be automatically merged. 

Author: ```suggestion
pull request
```

(we have this mistake multiple places, correct it...)

    - Write code to fix the bug
4. **Follow steps 4 and 5 in "[Checklist: I have a component request](#checklist-i-have-a-component-request)".**

## Setup environment and tooling for contribution

<!-- The following section is responsible for teaching people how to get setup with 
the repo such that they can get coding -->
This section will help you get going with Kirby such that you can begin contributing. 

### Clone and install the repository

<!-- how to clone and get going with the kirby repo-->
**Note:** make sure you have the right version of Node installed before installing the repository (see: [Node](#node)).

1. Clone the repository: 

``` 

git clone https://github.com/kirbydesign/designsystem
```

2. Install dependencies: 

``` 

npm install
```

3. Start the cookbook server:

``` 

npm start
```

You should now be able to open `http://localhost:4200` in a browser to access the cookbook.

### Node 

<!-- Which version should people use, and perhaps a recommendation on which tool people can use to manage versions. -->
[Node](https://nodejs.org/en/) is the runtime used for development. 

> It is recommended to use `v12` for development as we currently do not support node LTS (see issue: [#1443](https://github.com/kirbydesign/designsystem/issues/1443)). 

Reviewer 4: ```suggestion
The supported version of NodeJS can be found in the file `.tool-versions` in the repository root.
```

Author: Can't seem to find `.tool-versions` did it end up in https://github.com/kirbydesign/designsystem/blob/master/.nvmrc instead?

TSLint is used for linting which helps keep up the quality of the submitted code for Kirby. 

Scripts for running TSLint can be found in `package.json` . We however recommend that you install a TSLint plugin for your editor. It can be useful to enable "fix on save"-functionality such that simple errors are fixed automatically. 

**Useful resources:**

* [TSLint extension for VSCode](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin).
* [Enabling "fix on save" in VSCode](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin#auto-fix-and-auto-fix-on-save).

#### VSCode "workspace library execution"-issue

An issue might occur where TSLint will randomly disable for the current workspace and has to be enabled again using the command `TSLint: Manage workspace library execution`. 

To solve this: 

1. Open a typescript file in VScode. 
2. Open the command pallette (<kbd>ctrl/cmd</kbd> + <kbd>shift</kbd> + <kbd>p</kbd>) and enter: `typescript: select typescript version`

> 4. Select: `use workspace version 3.6.5`.

Author: Check up on whether this is still correct after #1318 has been solved.
